### PR TITLE
[TTAHUB-1199] Slow response while editing a goal with multiple objectives on the RTR page

### PR DESCRIPTION
**Pull request by GarrettEHill**
## Description of change
move the query to calculate onAR and onApprovedAR to data generation time instead to data fetch time.

## Issue(s)

* https://ocio-jira.acf.hhs.gov/browse/TTAHUB-1199

### Every PR

- [ ] Meets issue criteria
- [ ] JIRA ticket status updated
- [ ] Code is meaningfully tested
- [ ] Meets accessibility standards (WCAG 2.1 Levels A, AA)
- [ ] API Documentation updated
- [ ] Boundary diagram updated
- [ ] Logical Data Model updated
- [ ] [Architectural Decision Records](https://adr.github.io/) written for major infrastructure decisions

### Production Deploy

- [ ] Staging smoke test completed

### After merge/deploy

- [ ] Update JIRA ticket status

### CODE REVIEW

File: src/models/hooks/activityReportObjectiveFile.js [hunk 1]
```diff
@@ -0,0 +1,48 @@
+const recalculateOnAR = async (sequelize, instance, options) => {
+  let objectiveCondition;
+  if (options.hookMetadata === undefined
+    || options.hookMetadata === null) {
+    throw new Error('hookMetadata is required for hook to function');
```
Comment (GarrettEHill) [inline src/models/hooks/activityReportObjectiveFile.js:5]: look up value instead of throwing

**Review by kryswisnaskas — COMMENTED** [review]
This is not ready to be reviewed yet, right?

(so far I saw performance improve 👍 but objectives don't load in the edit form on the RTR)

**Comment by thewatermethod** [comment]
Fetching goals for activity reports generates this error (likely culprit to be found in the objective reducer)

Screen Shot 2022-12-08 at 11 04 19 AM

**Comment by thewatermethod** [comment]
Everything on the RTR looked good, once the above is resolved, I'd like to verify the "onApprovedAR" stuff in the hooks through the UI.

Still seeing save/submit be pretty slow on the RTR - although definitely better! I wonder if there are any indexes that can be created. I'm also wondering if some of the hooks can be moved to worker processes? Maybe that was already thought of and rejected for some reason. Regardless, nothing that needs to be a part of this particular change, which is some nice cleanup/improvements

**Comment by thewatermethod** [comment]
✅ Created objectives via the RTR. 
✅ Added those two objectives to a report, saved, viewed on the RTR. The correct "onApproved" & "onAnyReport" fields were returned via the API.
❌ Removed one, submitted report. Removed objective still shows "onAnyReport" in the API response despite also showing no "activityReports." (See screenshot)

Screen Shot 2022-12-12 at 10 44 51 AM

File: src/models/hooks/objectiveTopic.js [hunk 2]
```diff
@@ -1,6 +1,28 @@
...
+const autoPopulateOnAR = (sequelize, instance, options) => {
+  // eslint-disable-next-line no-prototype-builtins
+  if (instance.onAR === undefined
+    || instance.onAR === null) {
+    instance.set('onAR', false);
+    if (!options.fields.includes('onAR')) {
+      options.fields.push('onAR');
+    }
+  }
+};
```
Comment (kryswisnaskas) [inline src/models/hooks/objectiveTopic.js:13]: Should we also "unset" this flag if an objective is removed from a draft AR?

Right now, I am seeing a scenario where a new objective is added to an AR, the tables get updated with the "onAR" true, but then after removal, the true value persists. The removed objective shows in the RTR section.

**Comment by thewatermethod** [comment]
@GarrettEHill is this ready for another look?

**Comment by GarrettEHill** [comment]
Yes, please review

**Comment by thewatermethod** [comment]
I just started re-kicking the tires on this, but one think I noticed is that if you create a multi-grant goal from the recipient record (RTR, click add new goals, select 2 grants, enter goal and objective information, click save), the "goal review" page, that is, that page with the newly created goal in the grey boxes, shows a grey box for each grant. Meaning, if I pick two grants, enter one goal, click save and continue, I will see two grey boxes, each containing identical goal information. 
That isn't the case on main so something must a little tweaked with the roll up.

**Comment by thewatermethod** [comment]
It looks like also that the reduction of objectives returns the sequelize model instance for filesy (the following is from the same request to save goals. This errors because of the state that existing files are returned in when loaded from the API)

Screen Shot 2023-01-06 at 2 24 07 PM

**Comment by thewatermethod** [comment]
Sadly, this also means I can't conclude my testing, as the review section no longer loads with the files in this state.

Screen Shot 2023-01-06 at 2 30 05 PM

Screen Shot 2023-01-06 at 2 30 24 PM

**Comment by thewatermethod** [comment]
Still seeing some similar issues, fresh DB, latest branch. RTR worked good, no issues with doubled goals. 

On the RTR, created a goal for 1 recipient with 2 grants, and added an objective with 3 topics, 2 resources, and 2 files.
When I go to the AR and save the goal as is, with all relations intact, there seems to be some issues with what is returned from the goal service.

The UI is missing 1 resource and all the files:
Screen Shot 2023-01-12 at 9 20 28 AM

These first two screenshots are the request sent from the client (note the 2 files, 2 resources present)

Screen Shot 2023-01-12 at 9 19 39 AM

Screen Shot 2023-01-12 at 9 20 08 AM

This last is what is returned (note that only one resource is returned, and the file is in the wrong format for consumption)
Screen Shot 2023-01-12 at 9 20 20 AM

**Comment by thewatermethod** [comment]
You probably also want to merge ```main``` since it looks like there's a conflict with between this branch and ```src/services/goals.js```

**Comment by thewatermethod** [comment]
@GarrettEHill is this ready for another test?

**Comment by GarrettEHill** [comment]
@thewatermethod yes please

**Review by thewatermethod — APPROVED** [review]
My test went well! Looks good, works much more quickly. I would have someone else review before merging since the changes are substantial

**Review by AdamAdHocTeam — COMMENTED** [review]
Can you please provide comments for each of the SQL statements in the hooks.

If applicable add comments for each path of an IF ELSE.

I think this will help us know what's going on at a glance if we have to revisit.

File: src/migrations/20221116000000-rtr-goal-performance.js [hunk 3]
```diff
@@ -0,0 +1,210 @@
+module.exports = {
```
Comment (AdamAdHocTeam) [inline src/migrations/20221116000000-rtr-goal-performance.js:1]: Please add comments to this file. A description of overall function and each section were applicable.

File: src/models/helpers/hookMetadata.js [hunk 4]
```diff
@@ -0,0 +1,28 @@
+const hasHookMetadata = (options) => (options.hookMetadata !== undefined
```
Comment (AdamAdHocTeam) [inline src/models/helpers/hookMetadata.js:1]: What exactly is this file and how is it used?

Comment (AdamAdHocTeam) [reply src/models/helpers/hookMetadata.js:1]: comments

File: src/models/activityReportGoal.js [hunk 5]
```diff
@@ -1,6 +1,7 @@
 const { Model } = require('sequelize');
 const { CLOSE_SUSPEND_REASONS } = require('../constants');
 const { formatDate } = require('../lib/modelHelpers');
+const { afterDestroy } = require('./hooks/activityReportGoal');
```
Comment (AdamAdHocTeam) [inline src/models/activityReportGoal.js:4]: Looks like we are adding 'afterDestroy' to a bunch of models. Why was the decision made to update the various fields from this function as opposed to using an existing one like 'afterUpdate'?

File: src/models/hooks/activityReport.js [hunk 6]
```diff
@@ -324,17 +324,52 @@ const propagateApprovedStatus = async (sequelize, instance, options) => {
       }
 
       if (objectives && objectives.length > 0) {
```
Comment (AdamAdHocTeam) [inline src/models/hooks/activityReport.js:326]: A comment like 'when the report is approved mark all related objective models as onApproved' true.

File: src/models/hooks/activityReportObjectiveFile.js [hunk 7]
```diff
@@ -0,0 +1,62 @@
+const { getSingularOrPluralData } = require('../helpers/hookMetadata');
+
+const recalculateOnAR = async (sequelize, instance, options) => {
+  // check to see if objectiveId or objectiveIds is validly defined
+  // when defined a more efficient search can be used
+  const objectiveIds = getSingularOrPluralData(options, 'objectiveId', 'objectiveIds');
+
```
Comment (AdamAdHocTeam) [inline src/models/hooks/activityReportObjectiveFile.js:7]: Again would be nice to have a comment for each path of the IF ELSE that explains what we do for each.

File: src/models/activityReportGoal.js (hunk 5 above)
Comment (GarrettEHill) [reply src/models/activityReportGoal.js:4]: when a goal is removed from a report, we check to see if its still on any other report, then populate if a change needs to occur. The same goes for any other entity hanging off of a report.

**Comment by GarrettEHill** [comment]
https://github.com/HHS/Head-Start-TTADP/pull/1308
and
https://github.com/HHS/Head-Start-TTADP/pull/1206
 need to be applied before this change as not not break them, as this introduces DB changes

**Comment by GarrettEHill** [comment]
@thewatermethod I made changes to get tests to reach 75% threshold. Just want to give you a chance to see this before I merge.

**Review by thewatermethod — APPROVED** [review]
thanks for adding those tests

**Pull request merged**